expression/builtin_string.go

```diff
@@ -3627,3 +3630,158 @@ type loadFileFunctionClass struct {
 func (c *loadFileFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
 	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "load_file")
 }
+
+type weightStringPadding byte
+
+const (
+	// weightStringPaddingNone is used for WEIGHT_STRING(expr) if the expr is non-numeric.
+	weightStringPaddingNone weightStringPadding = 0xFF
+	// weightStringPaddingAsChar is used for WEIGHT_STRING(expr AS CHAR(x)) and the expr is non-numeric.
+	weightStringPaddingAsChar = 0x20
+	// weightStringPaddingAsBinary is used for WEIGHT_STRING(expr as BINARY(x)) and the expr is non-numeric.
+	weightStringPaddingAsBinary = 0x00
+	// weightStringPaddingNull is used for WEIGHT_STRING(expr [AS (CHAR|BINARY)]) if the expr is numeric.
+	weightStringPaddingNull = 0xFE
```

Comment:
On MySQL 8.0 this "NULL" transformation only applies to `CHAR`.

```sql
mysql> select weight_string(456789 as binary(4)), weight_string(3.1415926535e0 as binary(4)), weight_string(456789 as char(4)), weight_string(3.1415926535e0 as char(4))\G
*************************** 1. row ***************************
        weight_string(456789 as binary(4)): 4567
weight_string(3.1415926535e0 as binary(4)): 3.14
          weight_string(456789 as char(4)): NULL
  weight_string(3.1415926535e0 as char(4)): NULL
1 row in set, 2 warnings (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------+
| Level   | Code | Message                                             |
+---------+------+-----------------------------------------------------+
| Warning | 1292 | Truncated incorrect BINARY(4) value: '456789'       |
| Warning | 1292 | Truncated incorrect BINARY(4) value: '3.1415926535' |
+---------+------+-----------------------------------------------------+
2 rows in set (0.00 sec)
```

(The same for MariaDB.)

Reply:
Addressed, thanks!